// @polywrap/web3-config-bundle-js/build/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.bundle = void 0;
var uri_resolver_extensions_js_1 = require("@polywrap/uri-resolver-extensions-js");
// $start: bundle
var Sys = __importStar(require("@polywrap/sys-config-bundle-js"));
var EthWallet = __importStar(require("@polywrap/ethereum-wallet-js"));
exports.bundle = {
    concurrent: Sys.bundle.concurrent,
    http: Sys.bundle.http,
    ipfsHttpClient: Sys.bundle.ipfsHttpClient,
    ipfsResolver: Sys.bundle.ipfsResolver,
    ethereumWallet: {
        uri: "plugin/ethereum-wallet@1.0",
        package: EthWallet.plugin({
            connections: new EthWallet.Connections({
                networks: {
                    mainnet: new EthWallet.Connection({
                        provider: "https://mainnet.infura.io/v3/b00b2c2cc09c487685e9fb061256d6a6",
                    }),
                    goerli: new EthWallet.Connection({
                        provider: "https://goerli.infura.io/v3/b00b2c2cc09c487685e9fb061256d6a6",
                    }),
                },
            }),
        }),
        implements: ["wrapscan.io/polywrap/ethereum-wallet@1.0"],
        redirectFrom: ["wrapscan.io/polywrap/ethereum-wallet@1.0"],
    },
    ensTextRecordResolver: {
        uri: "ipfs/QmdYoDrXPxgjSoWuSWirWYxU5BLtpGVKd3z2GXKhW2VXLh",
        implements: [
            "wrapscan.io/polywrap/ens-text-record-uri-resolver@1.0",
            uri_resolver_extensions_js_1.ExtendableUriResolver.defaultExtInterfaceUris[0].uri,
        ],
        redirectFrom: ["wrapscan.io/polywrap/ens-text-record-uri-resolver@1.0"],
    },
    ensContenthashResolver: {
        uri: "wrapscan.io/polywrap/ens-contenthash-uri-resolver@1.0",
        implements: [uri_resolver_extensions_js_1.ExtendableUriResolver.defaultExtInterfaceUris[0].uri],
    },
    ensIpfsContenthashResolver: {
        uri: "wrapscan.io/polywrap/ens-ipfs-contenthash-uri-resolver@1.0",
        implements: [uri_resolver_extensions_js_1.ExtendableUriResolver.defaultExtInterfaceUris[0].uri],
    },
};
// $end
