// @polywrap/file-system-plugin-js/build/wrap/types.js
"use strict";
/// NOTE: This is an auto-generated file.
///       All modifications will be overwritten.
Object.defineProperty(exports, "__esModule", { value: true });
exports.EncodingEnum = void 0;
/// Env START ///
/// Env END ///
/// Objects START ///
/// Objects END ///
/// Enums START ///
var EncodingEnum;
(function (EncodingEnum) {
    EncodingEnum[EncodingEnum["ASCII"] = 0] = "ASCII";
    EncodingEnum[EncodingEnum["UTF8"] = 1] = "UTF8";
    EncodingEnum[EncodingEnum["UTF16LE"] = 2] = "UTF16LE";
    EncodingEnum[EncodingEnum["UCS2"] = 3] = "UCS2";
    EncodingEnum[EncodingEnum["BASE64"] = 4] = "BASE64";
    EncodingEnum[EncodingEnum["BASE64URL"] = 5] = "BASE64URL";
    EncodingEnum[EncodingEnum["LATIN1"] = 6] = "LATIN1";
    EncodingEnum[EncodingEnum["BINARY"] = 7] = "BINARY";
    EncodingEnum[EncodingEnum["HEX"] = 8] = "HEX";
})(EncodingEnum = exports.EncodingEnum || (exports.EncodingEnum = {}));
/// Enums END ///
/// Imported Objects START ///
/// Imported Objects END ///
/// Imported Modules START ///
/// Imported Modules END ///
