// @polywrap/client-config-builder-js/build/PolywrapClientConfigBuilder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PolywrapClientConfigBuilder = void 0;
var bundles_1 = require("./bundles");
var BaseClientConfigBuilder_1 = require("./BaseClientConfigBuilder");
var core_js_1 = require("@polywrap/core-js");
var uri_resolvers_js_1 = require("@polywrap/uri-resolvers-js");
var uri_resolver_extensions_js_1 = require("@polywrap/uri-resolver-extensions-js");
var PolywrapClientConfigBuilder = /** @class */ (function (_super) {
    __extends(PolywrapClientConfigBuilder, _super);
    // $start: ClientConfigBuilder-constructor
    /**
     * Instantiate a PolywrapClientConfigBuilder
     */
    function PolywrapClientConfigBuilder() {
        return _super.call(this) || this;
    }
    PolywrapClientConfigBuilder.prototype.addDefaults = function () {
        this.addBundle("sys");
        this.addBundle("web3");
        return this;
    };
    PolywrapClientConfigBuilder.prototype.addBundle = function (bundle) {
        return (0, bundles_1.addBundle)(bundle, this);
    };
    PolywrapClientConfigBuilder.prototype.build = function (options) {
        var resolver = options && "resolver" in options ? options.resolver : undefined;
        var resolutionResultCache = options && "resolutionResultCache" in options
            ? options.resolutionResultCache
            : undefined;
        return {
            envs: this.buildEnvs(),
            interfaces: this.buildInterfaces(),
            resolver: resolver !== null && resolver !== void 0 ? resolver : uri_resolvers_js_1.RecursiveResolver.from(uri_resolvers_js_1.RequestSynchronizerResolver.from(uri_resolvers_js_1.ResolutionResultCacheResolver.from(__spreadArray(__spreadArray([
                uri_resolvers_js_1.StaticResolver.from(__spreadArray(__spreadArray(__spreadArray([], __read(this.buildRedirects()), false), __read(this.buildWrappers()), false), __read(this.buildPackages()), false))
            ], __read(this._config.resolvers), false), [
                new uri_resolver_extensions_js_1.ExtendableUriResolver(),
            ], false), resolutionResultCache !== null && resolutionResultCache !== void 0 ? resolutionResultCache : new uri_resolvers_js_1.ResolutionResultCache()))),
        };
    };
    Object.defineProperty(PolywrapClientConfigBuilder.prototype, "config", {
        get: function () {
            return this._config;
        },
        enumerable: false,
        configurable: true
    });
    PolywrapClientConfigBuilder.prototype.buildEnvs = function () {
        var envs = new core_js_1.UriMap();
        for (var uri in this._config.envs) {
            envs.set(core_js_1.Uri.from(uri), this._config.envs[uri]);
        }
        return envs;
    };
    PolywrapClientConfigBuilder.prototype.buildInterfaces = function () {
        var interfaceImplementations = new core_js_1.UriMap();
        for (var uri in this._config.interfaces) {
            var uriImpls = __spreadArray([], __read(this._config.interfaces[uri]), false).map(function (x) {
                return core_js_1.Uri.from(x);
            });
            interfaceImplementations.set(core_js_1.Uri.from(uri), uriImpls);
        }
        return interfaceImplementations;
    };
    PolywrapClientConfigBuilder.prototype.buildRedirects = function () {
        var e_1, _a;
        var redirects = [];
        try {
            for (var _b = __values(Object.entries(this._config.redirects)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var _d = __read(_c.value, 2), uri = _d[0], redirect = _d[1];
                redirects.push({ from: core_js_1.Uri.from(uri), to: core_js_1.Uri.from(redirect) });
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return redirects;
    };
    PolywrapClientConfigBuilder.prototype.buildWrappers = function () {
        var e_2, _a;
        var wrappers = [];
        try {
            for (var _b = __values(Object.entries(this._config.wrappers)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var _d = __read(_c.value, 2), uri = _d[0], wrapper = _d[1];
                wrappers.push({ uri: core_js_1.Uri.from(uri), wrapper: wrapper });
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return wrappers;
    };
    PolywrapClientConfigBuilder.prototype.buildPackages = function () {
        var e_3, _a;
        var packages = [];
        try {
            for (var _b = __values(Object.entries(this._config.packages)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var _d = __read(_c.value, 2), uri = _d[0], wrapPackage = _d[1];
                packages.push({ uri: core_js_1.Uri.from(uri), package: wrapPackage });
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_3) throw e_3.error; }
        }
        return packages;
    };
    return PolywrapClientConfigBuilder;
}(BaseClientConfigBuilder_1.BaseClientConfigBuilder));
exports.PolywrapClientConfigBuilder = PolywrapClientConfigBuilder;
