// polywrap/build/commands/codegen.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.codegen = void 0;
var createLogger_1 = require("./utils/createLogger");
var lib_1 = require("../lib");
var ScriptCodeGenerator_1 = require("../lib/codegen/ScriptCodeGenerator");
var defaults_1 = require("../lib/defaults");
var watchProject_1 = require("../lib/watchProject");
var uri_1 = require("../lib/option-parsers/uri");
var client_js_1 = require("@polywrap/client-js");
var pathStr = lib_1.intlMsg.commands_codegen_options_o_path();
var defaultManifestStr = lib_1.defaultPolywrapManifestFiles.join(" | ");
exports.codegen = {
    setup: function (program) {
        program
            .command("codegen")
            .alias("g")
            .description(lib_1.intlMsg.commands_codegen_description())
            .option("-m, --manifest-file <".concat(pathStr, ">"), "".concat(lib_1.intlMsg.commands_codegen_options_m({
            default: defaultManifestStr,
        })))
            .option("-g, --codegen-dir <".concat(pathStr, ">"), "".concat(lib_1.intlMsg.commands_codegen_options_codegen({
            default: defaults_1.DEFAULT_CODEGEN_DIR,
        })))
            .option("-b, --bindgen <URI>", "".concat(lib_1.intlMsg.commands_codegen_options_b()))
            .option("-s, --script <".concat(pathStr, ">"), "".concat(lib_1.intlMsg.commands_codegen_options_s()))
            .option("-c, --client-config <".concat(lib_1.intlMsg.commands_common_options_configPath(), ">"), "".concat(lib_1.intlMsg.commands_common_options_config()))
            .option("--wrapper-envs <".concat(lib_1.intlMsg.commands_common_options_wrapperEnvsPath(), ">"), "".concat(lib_1.intlMsg.commands_common_options_wrapperEnvs()))
            .option("-w, --watch", "".concat(lib_1.intlMsg.commands_common_options_w()))
            .option("-v, --verbose", lib_1.intlMsg.commands_common_options_verbose())
            .option("-q, --quiet", lib_1.intlMsg.commands_common_options_quiet())
            .option("-l, --log-file [".concat(pathStr, "]"), "".concat(lib_1.intlMsg.commands_build_options_l()))
            .action(function (options) { return __awaiter(void 0, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, run({
                            manifestFile: (0, lib_1.parseManifestFileOption)(options.manifestFile, lib_1.defaultProjectManifestFiles),
                            codegenDir: (0, lib_1.parseDirOptionNoDefault)(options.codegenDir),
                            bindgen: options.bindgen || false,
                            script: (0, lib_1.parseCodegenScriptOption)(options.script),
                            clientConfig: options.clientConfig || false,
                            wrapperEnvs: options.wrapperEnvs || false,
                            verbose: options.verbose || false,
                            quiet: options.quiet || false,
                            logFile: (0, lib_1.parseLogFileOption)(options.logFile),
                            watch: options.watch || false,
                        })];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        }); });
    },
};
function run(options) {
    return __awaiter(this, void 0, void 0, function () {
        var manifestFile, clientConfig, wrapperEnvs, codegenDir, bindgen, script, verbose, quiet, logFile, watch, logger, bindgenUri, envs, configBuilder, client, project, schemaComposer, codeGenerator, execute, result;
        var _this = this;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    manifestFile = options.manifestFile, clientConfig = options.clientConfig, wrapperEnvs = options.wrapperEnvs, codegenDir = options.codegenDir, bindgen = options.bindgen, script = options.script, verbose = options.verbose, quiet = options.quiet, logFile = options.logFile, watch = options.watch;
                    logger = (0, createLogger_1.createLogger)({ verbose: verbose, quiet: quiet, logFile: logFile });
                    bindgenUri = (0, uri_1.parseUriOption)(bindgen);
                    return [4 /*yield*/, (0, lib_1.parseWrapperEnvsOption)(wrapperEnvs)];
                case 1:
                    envs = _a.sent();
                    return [4 /*yield*/, (0, lib_1.parseClientConfigOption)(clientConfig)];
                case 2:
                    configBuilder = _a.sent();
                    if (envs) {
                        configBuilder.addEnvs(envs);
                    }
                    client = new client_js_1.PolywrapClient(configBuilder.build());
                    return [4 /*yield*/, (0, lib_1.getProjectFromManifest)(manifestFile, logger)];
                case 3:
                    project = _a.sent();
                    if (!project) {
                        logger.error("".concat(lib_1.intlMsg.commands_codegen_project_load_error({
                            manifestPath: manifestFile,
                        })));
                        process.exit(1);
                    }
                    schemaComposer = new lib_1.SchemaComposer({
                        project: project,
                        client: client,
                    });
                    codeGenerator = script
                        ? new ScriptCodeGenerator_1.ScriptCodegenerator({
                            codegenDirAbs: codegenDir || undefined,
                            script: script,
                            schemaComposer: schemaComposer,
                            project: project,
                            omitHeader: false,
                            mustacheView: undefined,
                        })
                        : new lib_1.CodeGenerator({
                            codegenDirAbs: codegenDir || undefined,
                            schemaComposer: schemaComposer,
                            project: project,
                            bindgenUri: bindgenUri,
                        });
                    execute = function () { return __awaiter(_this, void 0, void 0, function () {
                        var result, err_1;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0:
                                    result = false;
                                    _a.label = 1;
                                case 1:
                                    _a.trys.push([1, 3, , 4]);
                                    return [4 /*yield*/, codeGenerator.generate()];
                                case 2:
                                    result = _a.sent();
                                    if (result) {
                                        logger.info("\uD83D\uDD25 ".concat(lib_1.intlMsg.commands_codegen_success(), " \uD83D\uDD25"));
                                    }
                                    return [3 /*break*/, 4];
                                case 3:
                                    err_1 = _a.sent();
                                    logger.error(err_1.message);
                                    return [2 /*return*/, false];
                                case 4: return [2 /*return*/, result];
                            }
                        });
                    }); };
                    if (!!watch) return [3 /*break*/, 5];
                    return [4 /*yield*/, execute()];
                case 4:
                    result = _a.sent();
                    if (!result) {
                        process.exit(1);
                    }
                    process.exit(0);
                    return [3 /*break*/, 7];
                case 5: return [4 /*yield*/, (0, watchProject_1.watchProject)({
                        execute: execute,
                        logger: logger,
                        project: project,
                        ignored: [codegenDir + "/**", project.getManifestDir() + "/**/wrap/**"],
                    })];
                case 6:
                    _a.sent();
                    _a.label = 7;
                case 7: return [2 /*return*/];
            }
        });
    });
}
