// polywrap/build/commands/infra.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.infra = exports.InfraActions = void 0;
var lib_1 = require("../lib");
var createLogger_1 = require("./utils/createLogger");
var path_1 = __importDefault(require("path"));
var commander_1 = require("commander");
var chalk_1 = __importDefault(require("chalk"));
var fs_1 = require("fs");
var InfraActions;
(function (InfraActions) {
    InfraActions["UP"] = "up";
    InfraActions["DOWN"] = "down";
    InfraActions["VARS"] = "vars";
    InfraActions["CONFIG"] = "config";
})(InfraActions = exports.InfraActions || (exports.InfraActions = {}));
var DEFAULT_MODULES_PATH = path_1.default.join(__dirname, "..", "lib", "defaults", "infra-modules");
var defaultManifestStr = lib_1.defaultInfraManifestFiles.join(" | ");
var pathStr = lib_1.intlMsg.commands_infra_options_m_path();
var moduleNameStr = lib_1.intlMsg.commands_infra_options_o_module();
var argumentsDescription = "\n  ".concat(lib_1.intlMsg.commands_infra_actions_subtitle(), "\n  ").concat(chalk_1.default.bold(InfraActions.UP), "      ").concat(lib_1.intlMsg.commands_infra_command_up(), "\n  ").concat(chalk_1.default.bold(InfraActions.DOWN), "    ").concat(lib_1.intlMsg.commands_infra_command_down(), "\n  ").concat(chalk_1.default.bold(InfraActions.CONFIG), "   ").concat(lib_1.intlMsg.commands_infra_command_config(), "\n  ").concat(chalk_1.default.bold(InfraActions.VARS), "     ").concat(lib_1.intlMsg.commands_infra_command_vars(), "\n");
exports.infra = {
    setup: function (program) {
        program
            .command("infra")
            .alias("i")
            .description(lib_1.intlMsg.commands_infra_description())
            .usage("<action> [options]")
            .addArgument(new commander_1.Argument("<action>", argumentsDescription).choices([
            InfraActions.UP,
            InfraActions.DOWN,
            InfraActions.VARS,
            InfraActions.CONFIG,
        ]))
            .showHelpAfterError(true)
            .option("-m, --manifest-file  <".concat(pathStr, ">"), lib_1.intlMsg.commands_infra_options_m({
            default: defaultManifestStr,
        }))
            .option("-o, --modules <".concat(moduleNameStr, "...>"), lib_1.intlMsg.commands_infra_options_o())
            .option("-v, --verbose", lib_1.intlMsg.commands_common_options_verbose())
            .option("-q, --quiet", lib_1.intlMsg.commands_common_options_quiet())
            .option("-l, --log-file [".concat(pathStr, "]"), "".concat(lib_1.intlMsg.commands_build_options_l()))
            .action(function (action, options) { return __awaiter(void 0, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, run(action, {
                            manifestFile: options.manifestFile || false,
                            modules: options.modules || false,
                            verbose: options.verbose || false,
                            quiet: options.quiet || false,
                            logFile: (0, lib_1.parseLogFileOption)(options.logFile),
                        })];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        }); });
    },
};
var tip = "Tip: If no infra manifest is specified, a default module should be specified using the '--modules' option.\n\nDefault Modules: \n".concat((0, fs_1.readdirSync)(DEFAULT_MODULES_PATH)
    .map(function (m) { return "\n- ".concat(m); })
    .join(""), "\n\nExample: 'polywrap infra up --modules=eth-ens-ipfs'.");
function run(action, options) {
    return __awaiter(this, void 0, void 0, function () {
        var modules, verbose, quiet, manifestFile, logFile, logger, modulesArray, manifest, manifestPath, infraManifest, e_1, infra, filteredModules, errorMsg_1, errorMsg, _a, _b, _c, _d, _e;
        return __generator(this, function (_f) {
            switch (_f.label) {
                case 0:
                    modules = options.modules, verbose = options.verbose, quiet = options.quiet, manifestFile = options.manifestFile, logFile = options.logFile;
                    logger = (0, createLogger_1.createLogger)({ verbose: verbose, quiet: quiet, logFile: logFile });
                    modulesArray = [];
                    if (modules) {
                        modules.forEach(function (x) {
                            return modulesArray.push.apply(modulesArray, __spreadArray([], __read((x.includes(",") ? x.split(",") : [x])), false));
                        });
                    }
                    manifest = manifestFile
                        ? [manifestFile]
                        : lib_1.defaultInfraManifestFiles;
                    manifestPath = (0, lib_1.resolvePathIfExists)(manifest);
                    if (!manifestPath) return [3 /*break*/, 5];
                    _f.label = 1;
                case 1:
                    _f.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, (0, lib_1.loadInfraManifest)(manifestPath, logger)];
                case 2:
                    infraManifest = _f.sent();
                    return [3 /*break*/, 4];
                case 3:
                    e_1 = _f.sent();
                    if (!modulesArray.length) {
                        throw new Error("".concat(e_1.message, "\n\n").concat(tip));
                    }
                    else {
                        throw new Error(e_1.message);
                    }
                    return [3 /*break*/, 4];
                case 4: return [3 /*break*/, 6];
                case 5:
                    if (!modulesArray.length) {
                        throw new Error(tip);
                    }
                    _f.label = 6;
                case 6:
                    infra = new lib_1.Infra({
                        rootDir: manifestPath ? path_1.default.dirname(manifestPath) : process.cwd(),
                        modulesToUse: modulesArray,
                        infraManifest: infraManifest,
                        defaultInfraModulesPath: DEFAULT_MODULES_PATH,
                        logger: logger,
                    });
                    filteredModules = infra.getFilteredModules();
                    if (!filteredModules.length) {
                        if (modules) {
                            errorMsg_1 = lib_1.intlMsg.commands_infra_error_noModulesMatch({
                                modules: modules.join(", "),
                            });
                            logger.error(errorMsg_1);
                            return [2 /*return*/];
                        }
                        errorMsg = lib_1.intlMsg.commands_infra_error_noModulesDeclared();
                        logger.error(errorMsg);
                        return [2 /*return*/];
                    }
                    logger.info("".concat(lib_1.intlMsg.commands_infra_modulesUsed_text(), ": ").concat(filteredModules
                        .map(function (f) { return "\n- ".concat(f.name); })
                        .join(""), "\n"));
                    _a = action;
                    switch (_a) {
                        case InfraActions.UP: return [3 /*break*/, 7];
                        case InfraActions.DOWN: return [3 /*break*/, 9];
                        case InfraActions.VARS: return [3 /*break*/, 11];
                        case InfraActions.CONFIG: return [3 /*break*/, 13];
                    }
                    return [3 /*break*/, 15];
                case 7: return [4 /*yield*/, infra.up()];
                case 8:
                    _f.sent();
                    return [3 /*break*/, 16];
                case 9: return [4 /*yield*/, infra.down()];
                case 10:
                    _f.sent();
                    return [3 /*break*/, 16];
                case 11:
                    _c = (_b = logger).info;
                    _e = (_d = JSON).stringify;
                    return [4 /*yield*/, infra.getVars()];
                case 12:
                    _c.apply(_b, [_e.apply(_d, [_f.sent(), null, 2])]);
                    return [3 /*break*/, 16];
                case 13: 
                // Calling "docker-compose config" will log to console, so we don't have to
                return [4 /*yield*/, infra.config()];
                case 14:
                    // Calling "docker-compose config" will log to console, so we don't have to
                    _f.sent();
                    return [3 /*break*/, 16];
                case 15: throw Error(lib_1.intlMsg.commands_infra_error_never());
                case 16:
                    process.exit(0);
                    return [2 /*return*/];
            }
        });
    });
}
