// polywrap/build/commands/test.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.test = void 0;
var lib_1 = require("../lib");
var createLogger_1 = require("./utils/createLogger");
var core_js_1 = require("@polywrap/core-js");
var path_1 = __importDefault(require("path"));
var yaml_1 = __importDefault(require("yaml"));
var fs_1 = __importDefault(require("fs"));
var defaultManifestStr = lib_1.defaultWorkflowManifestFiles.join(" | ");
var pathStr = lib_1.intlMsg.commands_test_options_m_path();
exports.test = {
    setup: function (program) {
        program
            .command("test")
            .alias("t")
            .description(lib_1.intlMsg.commands_test_description())
            .option("-m, --manifest-file  <".concat(pathStr, ">"), lib_1.intlMsg.commands_test_options_m({
            default: defaultManifestStr,
        }))
            .option("-c, --client-config <".concat(lib_1.intlMsg.commands_common_options_configPath(), ">"), "".concat(lib_1.intlMsg.commands_common_options_config()))
            .option("--wrapper-envs <".concat(lib_1.intlMsg.commands_common_options_wrapperEnvsPath(), ">"), "".concat(lib_1.intlMsg.commands_common_options_wrapperEnvs()))
            .option("-o, --output-file <".concat(lib_1.intlMsg.commands_test_options_outputFilePath(), ">"), "".concat(lib_1.intlMsg.commands_test_options_outputFile()))
            .option("-j, --jobs <".concat(lib_1.intlMsg.commands_test_options_jobIds(), "...>"), lib_1.intlMsg.commands_test_options_jobs())
            .option("-v, --verbose", lib_1.intlMsg.commands_common_options_verbose())
            .option("-q, --quiet", lib_1.intlMsg.commands_common_options_quiet())
            .option("-l, --log-file [".concat(pathStr, "]"), "".concat(lib_1.intlMsg.commands_build_options_l()))
            .action(function (options) { return __awaiter(void 0, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, _run({
                            manifestFile: (0, lib_1.parseManifestFileOption)(options.manifestFile, lib_1.defaultWorkflowManifestFiles),
                            clientConfig: options.clientConfig || false,
                            wrapperEnvs: options.wrapperEnvs || false,
                            outputFile: options.outputFile
                                ? (0, lib_1.parseWorkflowOutputFilePathOption)(options.outputFile)
                                : false,
                            jobs: options.jobs || false,
                            validationScript: options.validationScript || false,
                            verbose: options.verbose || false,
                            quiet: options.quiet || false,
                            logFile: (0, lib_1.parseLogFileOption)(options.logFile),
                        })];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        }); });
    },
};
var _run = function (options) { return __awaiter(void 0, void 0, void 0, function () {
    var manifestFile, clientConfig, wrapperEnvs, outputFile, jobs, verbose, quiet, logFile, logger, envs, configBuilder, manifestPath, workflow, jobsArray, validationScript, workflowOutput, onExecution, jobRunner, outputFileExt, printableOutput, isFailed;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                manifestFile = options.manifestFile, clientConfig = options.clientConfig, wrapperEnvs = options.wrapperEnvs, outputFile = options.outputFile, jobs = options.jobs, verbose = options.verbose, quiet = options.quiet, logFile = options.logFile;
                logger = (0, createLogger_1.createLogger)({ verbose: verbose, quiet: quiet, logFile: logFile });
                return [4 /*yield*/, (0, lib_1.parseWrapperEnvsOption)(wrapperEnvs)];
            case 1:
                envs = _a.sent();
                return [4 /*yield*/, (0, lib_1.parseClientConfigOption)(clientConfig)];
            case 2:
                configBuilder = _a.sent();
                if (envs) {
                    configBuilder.addEnvs(envs);
                }
                manifestPath = path_1.default.resolve(manifestFile);
                return [4 /*yield*/, (0, lib_1.loadWorkflowManifest)(manifestPath, logger)];
            case 3:
                workflow = _a.sent();
                (0, lib_1.validateJobNames)(workflow.jobs);
                jobsArray = [];
                if (jobs) {
                    jobs.forEach(function (x) {
                        return jobsArray.push.apply(jobsArray, __spreadArray([], __read((x.includes(",") ? x.split(",") : [x])), false));
                    });
                }
                else {
                    jobsArray.push.apply(jobsArray, __spreadArray([], __read(Object.keys(workflow.jobs)), false));
                }
                validationScript = workflow.validation
                    ? (0, lib_1.loadValidationScript)(manifestPath, workflow.validation)
                    : undefined;
                workflowOutput = [];
                onExecution = function (id, jobResult) {
                    var data = jobResult.data, error = jobResult.error, status = jobResult.status;
                    var output = {
                        id: id,
                        status: status,
                        data: data,
                        error: error,
                        validation: {
                            status: lib_1.Status.SKIPPED,
                        },
                    };
                    if (validationScript) {
                        (0, lib_1.validateOutput)(output, validationScript, logger);
                    }
                    if (!quiet) {
                        (0, lib_1.printJobOutput)(output);
                    }
                    workflowOutput.push(output);
                };
                jobRunner = new lib_1.JobRunner(configBuilder, onExecution);
                return [4 /*yield*/, jobRunner.run(workflow.jobs, jobsArray)];
            case 4:
                _a.sent();
                if (outputFile) {
                    outputFileExt = path_1.default.extname(outputFile).substring(1);
                    if (!outputFileExt)
                        throw new Error("Require output file extension");
                    printableOutput = workflowOutput.map(function (o) {
                        var _a;
                        return (__assign(__assign({}, o), { error: (_a = o.error) === null || _a === void 0 ? void 0 : _a.message }));
                    });
                    switch (outputFileExt) {
                        case "yaml":
                        case "yml":
                            fs_1.default.writeFileSync(outputFile, yaml_1.default.stringify(printableOutput, null, 2));
                            break;
                        case "json":
                            fs_1.default.writeFileSync(outputFile, JSON.stringify(printableOutput, core_js_1.typesHandler, 2));
                            break;
                        default:
                            throw new Error(lib_1.intlMsg.commands_test_error_unsupportedOutputFileExt({
                                outputFileExt: outputFileExt,
                            }));
                    }
                }
                isFailed = function (val) {
                    return val.status === lib_1.Status.FAILED &&
                        (val.validation.status === lib_1.Status.SKIPPED ||
                            val.validation.status === lib_1.Status.FAILED);
                };
                if (workflowOutput.some(isFailed)) {
                    process.exit(1);
                }
                return [2 /*return*/];
        }
    });
}); };
