// @polywrap/concurrent-plugin-js/build/wrap/types.js
"use strict";
/// NOTE: This is an auto-generated file.
///       All modifications will be overwritten.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReturnWhenEnum = exports.TaskStatusEnum = void 0;
/// Objects END ///
/// Enums START ///
var TaskStatusEnum;
(function (TaskStatusEnum) {
    TaskStatusEnum[TaskStatusEnum["PENDING"] = 0] = "PENDING";
    TaskStatusEnum[TaskStatusEnum["RUNNING"] = 1] = "RUNNING";
    TaskStatusEnum[TaskStatusEnum["COMPLETED"] = 2] = "COMPLETED";
    TaskStatusEnum[TaskStatusEnum["CANCELLED"] = 3] = "CANCELLED";
    TaskStatusEnum[TaskStatusEnum["FAILED"] = 4] = "FAILED";
})(TaskStatusEnum = exports.TaskStatusEnum || (exports.TaskStatusEnum = {}));
var ReturnWhenEnum;
(function (ReturnWhenEnum) {
    ReturnWhenEnum[ReturnWhenEnum["FIRST_COMPLETED"] = 0] = "FIRST_COMPLETED";
    ReturnWhenEnum[ReturnWhenEnum["ANY_COMPLETED"] = 1] = "ANY_COMPLETED";
    ReturnWhenEnum[ReturnWhenEnum["ALL_COMPLETED"] = 2] = "ALL_COMPLETED";
})(ReturnWhenEnum = exports.ReturnWhenEnum || (exports.ReturnWhenEnum = {}));
/// Enums END ///
/// Imported Objects START ///
/// Imported Objects END ///
/// Imported Modules START ///
/// Imported Modules END ///
