// @polywrap/schema-bind/build/bindings/golang/transforms/moduleNeedsTypes.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.moduleNeedsTypes = void 0;
/* eslint-disable @typescript-eslint/naming-convention */
var types_1 = require("../types");
function moduleNeedsTypes() {
    var state = {};
    return {
        enter: {
            Abi: function (abi) {
                // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                // @ts-ignore
                state.importedTypes = abi._importedTypes;
                return abi;
            },
            ModuleDefinition: function (def) {
                state.moduleDefinition = def;
                state.needsTypes = false;
                return def;
            },
            MethodDefinition: function (def) {
                var _a;
                if (def.arguments && def.arguments.length > 0) {
                    state.needsTypes = true;
                }
                if (def.return) {
                    var returnType = def.return.type;
                    if (!(0, types_1.isBaseType)(returnType) &&
                        !(0, types_1.isBuiltInType)(returnType) &&
                        !((_a = state.importedTypes) === null || _a === void 0 ? void 0 : _a.has(returnType))) {
                        state.needsTypes = true;
                    }
                }
                return def;
            },
        },
        leave: {
            ModuleDefinition: function (def) {
                var needsTypes = state.needsTypes;
                state.moduleDefinition = undefined;
                state.needsTypes = undefined;
                return __assign(__assign({}, def), { needsTypes: needsTypes });
            },
        },
    };
}
exports.moduleNeedsTypes = moduleNeedsTypes;
