// @polywrap/polywrap-manifest-types-js/build/formats/polywrap.docs/index.js
"use strict";
/* eslint-disable */
/**
 * This file was automatically generated by scripts/manifest/index-ts.mustache.
 * DO NOT MODIFY IT BY HAND. Instead, modify scripts/manifest/index-ts.mustache,
 * and run node ./scripts/manifest/generateFormatTypes.js to regenerate this file.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateDocsManifest = exports.deserializeDocsManifest = exports.migrateDocsManifest = exports.latestDocsManifestFormat = exports.DocsManifestSchemaFiles = exports.DocsManifestFormats = void 0;
var DocsManifestFormats;
(function (DocsManifestFormats) {
    // NOTE: Patch fix for backwards compatability
    DocsManifestFormats["v0.1"] = "0.1";
    DocsManifestFormats["v0.1.0"] = "0.1.0";
})(DocsManifestFormats = exports.DocsManifestFormats || (exports.DocsManifestFormats = {}));
exports.DocsManifestSchemaFiles = {
    // NOTE: Patch fix for backwards compatability
    "0.1": "formats/polywrap.docs/0.1.0.json",
    "0.1.0": "formats/polywrap.docs/0.1.0.json",
};
exports.latestDocsManifestFormat = DocsManifestFormats["v0.1.0"];
var migrate_1 = require("./migrate");
Object.defineProperty(exports, "migrateDocsManifest", { enumerable: true, get: function () { return migrate_1.migrateDocsManifest; } });
var deserialize_1 = require("./deserialize");
Object.defineProperty(exports, "deserializeDocsManifest", { enumerable: true, get: function () { return deserialize_1.deserializeDocsManifest; } });
var validate_1 = require("./validate");
Object.defineProperty(exports, "validateDocsManifest", { enumerable: true, get: function () { return validate_1.validateDocsManifest; } });
