// @polywrap/uri-resolvers-js/build/cache/WrapperCacheResolver.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WrapperCacheResolver = void 0;
var helpers_1 = require("../helpers");
var core_js_1 = require("@polywrap/core-js");
// $start: WrapperCacheResolver
/**
 * An IUriResolver implementation that caches wrappers once they are resolved.
 * As it is a wrapper cache resolver, URI and package caching is outside of the scope for this resolver
 * and can be achieved through other resolvers if necessary.
 * The WrapperCacheResolver wraps an IUriResolver implementation and delegates resolution to it.
 * */
var WrapperCacheResolver = /** @class */ (function () {
    // $start: WrapperCacheResolver-constructor
    /**
     * Creates a WrapperCacheResolver
     *
     * @param _innerResolver - a resolver to delegate resolution to
     * @param _cache - a wrapper cache
     * */
    function WrapperCacheResolver(_innerResolver, _cache) {
        this._innerResolver = _innerResolver;
        this._cache = _cache;
    }
    // $start: WrapperCacheResolver-from
    /**
     * Creates a WrapperCacheResolver from a resolver-like object
     *
     * @param innerResolver - a resolver-like item to delegate resolution to
     * @param cache - a wrapper cache
     * @param options - control wrapper manifest deserialization
     *
     * @returns a WrapperCacheResolver
     * */
    WrapperCacheResolver.from = function (innerResolver, cache) {
        return new WrapperCacheResolver(helpers_1.UriResolver.from(innerResolver), cache);
    };
    // $start: WrapperCacheResolver-tryResolveUri
    /**
     * Resolve a URI to a wrap package, a wrapper, or a URI.
     * If successful, cache the result.
     *
     * @param uri - the URI to resolve
     * @param client - a CoreClient instance that may be used to invoke a wrapper that implements the UriResolver interface
     * @param resolutionContext - the current URI resolution context
     * @returns A Promise with a Result containing either a wrap package, a wrapper, or a URI if successful
     */
    WrapperCacheResolver.prototype.tryResolveUri = function (uri, client, resolutionContext) {
        return __awaiter(this, void 0, void 0, function () {
            var wrapper, result_1, subContext, result, wrapper_1, resolutionPath, resolutionPath_1, resolutionPath_1_1, uri_1, e_1_1;
            var e_1, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this._cache.get(uri)];
                    case 1:
                        wrapper = _b.sent();
                        // Return from cache if available
                        if (wrapper) {
                            result_1 = core_js_1.UriResolutionResult.ok(uri, wrapper);
                            resolutionContext.trackStep({
                                sourceUri: uri,
                                result: result_1,
                                description: "WrapperCacheResolver (Cache)",
                            });
                            return [2 /*return*/, result_1];
                        }
                        subContext = resolutionContext.createSubHistoryContext();
                        return [4 /*yield*/, this._innerResolver.tryResolveUri(uri, client, subContext)];
                    case 2:
                        result = _b.sent();
                        if (!(result.ok && result.value.type === "wrapper")) return [3 /*break*/, 12];
                        wrapper_1 = result.value.wrapper;
                        resolutionPath = subContext.getResolutionPath();
                        _b.label = 3;
                    case 3:
                        _b.trys.push([3, 8, 9, 10]);
                        resolutionPath_1 = __values(resolutionPath), resolutionPath_1_1 = resolutionPath_1.next();
                        _b.label = 4;
                    case 4:
                        if (!!resolutionPath_1_1.done) return [3 /*break*/, 7];
                        uri_1 = resolutionPath_1_1.value;
                        return [4 /*yield*/, this._cache.set(uri_1, wrapper_1)];
                    case 5:
                        _b.sent();
                        _b.label = 6;
                    case 6:
                        resolutionPath_1_1 = resolutionPath_1.next();
                        return [3 /*break*/, 4];
                    case 7: return [3 /*break*/, 10];
                    case 8:
                        e_1_1 = _b.sent();
                        e_1 = { error: e_1_1 };
                        return [3 /*break*/, 10];
                    case 9:
                        try {
                            if (resolutionPath_1_1 && !resolutionPath_1_1.done && (_a = resolutionPath_1.return)) _a.call(resolutionPath_1);
                        }
                        finally { if (e_1) throw e_1.error; }
                        return [7 /*endfinally*/];
                    case 10: return [4 /*yield*/, this._cache.set(result.value.uri, wrapper_1)];
                    case 11:
                        _b.sent();
                        _b.label = 12;
                    case 12:
                        resolutionContext.trackStep({
                            sourceUri: uri,
                            result: result,
                            subHistory: subContext.getHistory(),
                            description: "WrapperCacheResolver",
                        });
                        return [2 /*return*/, result];
                }
            });
        });
    };
    return WrapperCacheResolver;
}());
exports.WrapperCacheResolver = WrapperCacheResolver;
